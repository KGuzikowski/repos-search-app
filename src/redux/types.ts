import { ThunkAction } from 'redux-thunk'
import { Action } from 'redux'

export type AppThunk<State, ReturnType = void> = ThunkAction<ReturnType, State, unknown, Action<string>>