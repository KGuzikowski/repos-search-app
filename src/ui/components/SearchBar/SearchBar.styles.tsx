import styled from 'styled-components'
import { secondaryColor } from '../../style-variables'
import { ReactComponent as Icon } from './search.svg'

export const Input = styled.input`
    padding 15px 20px 15px 60px;
    outline: none;
    box-sizing: border-box;
    font-size: 16px;
    border-color: #ccc;
    width: 100%;
    border: 1px solid #ccc;
    border-radius: 6px;

    &:focus {
        border-color: ${secondaryColor};
    }
`

export const Form = styled.form`
    margin-top: 20px;
    margin-bottom: 50px;
    width: 600px;
    
    @media only screen and (max-width: 700px) {
        width: 100%;
        margin-top: 20px;
        margin-bottom: 0px;
    }
`

export const SearchIcon = styled(Icon)`
    position: absolute;
    padding-top: 9px;
    padding-left: 12px;
    height: 35px;
    width: 35px;
`