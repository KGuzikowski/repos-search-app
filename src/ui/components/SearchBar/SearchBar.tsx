import React, { useState, useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from '../../../redux/rootReducer'
import { Input, Form, SearchIcon } from './SearchBar.styles'
import { secondaryColor } from '../../style-variables'
import { fetchReposStartAsync } from '../../../redux/repos/repo.actions'
import useDebounce from '../../../utils/useDebounce'

const SearchBar = () => {
    const name = useSelector((state: RootState) => state.repos.name)
    const [iconFill, setIconFill] = useState('#ccc')
    const [repoName, setRepoName] = useState(name)
    const dispatch = useDispatch()
    const debouncedRepoName = useDebounce(repoName, 500)
    
    useEffect(() => {
        if(debouncedRepoName) {
            dispatch(fetchReposStartAsync(debouncedRepoName))
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [debouncedRepoName])

    const handleSubmit = (e: React.FormEvent) => {
        e.preventDefault()
        dispatch(fetchReposStartAsync(repoName))
    }

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setRepoName(e.target.value)
    }

    return (
        <Form onSubmit={e => handleSubmit(e)}>
            <SearchIcon fill={iconFill} />
            <Input type='text' placeholder='Repository name' value={repoName} onChange={e => handleChange(e)} onFocus={() => setIconFill(secondaryColor)} onBlur={() => setIconFill('#ccc')}/>
        </Form>
    )
}

export default SearchBar