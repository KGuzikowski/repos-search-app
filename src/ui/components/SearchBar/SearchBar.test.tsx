import React from 'react'
import { create } from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import SearchBar from './SearchBar'
import { reposStateType } from '../../../redux/repos/repo.types'

it('renders SearchBar component correctly', () => {
    const repo = {
        id: 1,
        name: 'repo',
        owner: { id: 1, name: 'Karol' },
        stars: 2,
        createdAt: '2019-05-12'
    }

    const INITIAL_STATE: reposStateType = {
        repos: [repo],
        isFetching: false,
        errorMessage: '',
        name: 'react',
        pagination: 5,
        page: 1,
    }

    const mockStore = configureStore()
    const store = mockStore(INITIAL_STATE)

    const component = create(<Provider store={store}><SearchBar /></Provider>)
    expect(component.toJSON()).toMatchSnapshot()
})