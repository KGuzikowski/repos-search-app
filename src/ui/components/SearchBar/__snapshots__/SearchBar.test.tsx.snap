// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SearchBar component correctly 1`] = `
<form
  className="sc-AxirZ kHBfiv"
  onSubmit={[Function]}
>
  <svg
    className="sc-AxiKw kolHka"
    fill="#ccc"
  >
    search.svg
  </svg>
  <input
    className="sc-AxjAm gbUlxN"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Repository name"
    type="text"
  />
</form>
`;
